include/ctpg/ctpg.hpp

```diff
@@ -1820,13 +1890,26 @@ class parser<
         return parse(parse_options{}, buffer, error_stream);
     }
 
+    template<typename Context, typename Buffer, typename ErrorStream>
+    constexpr std::optional<root_value_type> cparse(Context&& context, const Buffer& buffer, ErrorStream& error_stream) const
+    {
+        return cparse(std::forward<Context>(context), parse_options{}, buffer, error_stream);
+    }
+
     template<typename Buffer, typename ErrorStream>
     constexpr std::optional<root_value_type> parse(parse_options options, const Buffer& buffer, ErrorStream& error_stream) const
+    {
+        return cparse(no_type{}, options, buffer, error_stream);
+    }
+
+    template<typename Context, typename Buffer, typename ErrorStream>
+    constexpr std::optional<root_value_type> cparse(Context&& context, parse_options options, const Buffer& buffer, ErrorStream& error_stream) const
     {
         detail::parser_value_stack_type_t<Buffer, empty_rules_count, value_variant_type> value_stack{};
         detail::parse_table_cursor_stack_type_t<Buffer, empty_rules_count> cursor_stack{};
 
-        detail::parse_state ps(cursor_stack, value_stack, error_stream, options, buffer.begin(), buffer.end());
+        detail::value_reductors<Context&&, value_variant_type, rule_tuple_type, rule_count> reductors(std::forward<Context>(context), rule_tuple);
```

Comment:
This is inconsistent. What owns the context? Can you comment on this?

Reply:
The idea was that I store a reference to the context inside `value_reductors` so that `invoke` doesn't require the context to be passed as its argument. And parser always works with references to the context (my previous implementation with passing by value didn't work as I expected, so, I've decided to use references instead).